### PR TITLE
CLID-197: Add TargetRepo, TargetTag to additionalImages

### PR DESCRIPTION
**Pull request by dorzel**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Github / Jira issue: 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code Improvements (Refactoring, Performance, CI upgrades, etc)
- [ ] Internal repo assets (diagrams / docs on github repo)
- [ ] This change requires a documentation update on openshift docs

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

## Expected Outcome
Please describe the outcome expected from the tests.

## Summary by CodeRabbit

* **New Features**
  * Support for specifying target repository and tag overrides when resolving image names.

* **Improvements**
  * Centralized path-building and normalization with clearer validation and more descriptive error messages.
  * Preserves and surfaces image origin information and improves handling of tag/digest edge cases.

* **Bug Fixes**
  * Added early checks to ensure source/destination paths are determinable and consolidated duplicate-detection logic.

* **Tests**
  * Expanded coverage for targetRepo/targetTag scenarios, deletion flows, and path-nesting behavior.

### CODE REVIEW

File: internal/pkg/additional/local_stored_collector.go [hunk 1]
```diff
@@ -51,68 +51,37 @@ func (o LocalStorageCollector) AdditionalImagesCollector(ctx context.Context) ([
...
+		targetTag := imgSpec.Tag
+		if img.TargetTag != "" {
+			targetTag = img.TargetTag
+		}
+
+		switch {
+		case o.Opts.IsMirrorToDisk(), o.Opts.IsMirrorToMirror():
+			tmpSrc, tmpDest = o.buildMirrorToDiskPaths(img, imgSpec, targetRepo, targetTag)
+		case o.Opts.IsDiskToMirror():
+			tmpSrc, tmpDest = o.buildDiskToMirrorPaths(img, imgSpec, targetRepo, targetTag)
```
Comment (aguidirh) [inline internal/pkg/additional/local_stored_collector.go:85]: I like this refactoring, makes it easier to read and understand. Thanks

File: internal/pkg/api/v2alpha1/type_config.go [hunk 2]
```diff
@@ -269,6 +275,30 @@ type Image struct {
...
+	//     target-name    = path-component
+	//     namespace      = path-component ['/' path-component]*
+	//     path-component = alpha-numeric [separator alpha-numeric]*
+	//     alpha-numeric  = /[a-z0-9]+/
+	//     separator      = /[_.]|__|[-]*/
+	TargetRepo string `json:"targetRepo,omitempty"`
+	// TargetTag is the tag the image will be mirrored with. If unset,
+	// the image will be mirrored with the provided tag in the Name
+	// field or a tag calculated from the partial digest.
+	TargetTag string `json:"targetTag,omitempty"`
```
Comment (aguidirh) [inline internal/pkg/api/v2alpha1/type_config.go:292]: I'm wondering if now that Image struct is becoming more specific for additionalImages if we should have a dedicated struct for BlockedImages. Sharing the same struct betweem them could bring problems in future.

File: internal/pkg/additional/local_stored_collector_test.go [hunk 3]
```diff
@@ -250,6 +250,332 @@ func TestAdditionalImageCollector(t *testing.T) {
 	})
 }
 
+func TestAdditionalImageCollectorWithTargetRepoAndTag(t *testing.T) {
```
Comment (aguidirh) [inline internal/pkg/additional/local_stored_collector_test.go:255]: As the code on these tests are similar, the code could be reused with table driven tests.

File: internal/pkg/api/v2alpha1/type_config_test.go [hunk 4]
```diff
@@ -33,8 +37,278 @@ func TestIsValidPathComponent(t *testing.T) {
...
 		t.Run(test.name, func(t *testing.T) {
 			result := IsValidPathComponent(test.targetCatalog)
-			if result != test.expectedResult {
-				t.Errorf("Expected %v, but got %v", test.expectedResult, result)
+			assert.Equal(t, test.expectedResult, result)
+		})
+	}
+}
+
+func TestGetUniqueNameWithTarget(t *testing.T) {
```
Comment (aguidirh) [inline internal/pkg/api/v2alpha1/type_config_test.go:45]: +1 for the use of table driven tests here.

File: internal/pkg/additional/local_stored_collector_test.go (hunk 3 above)
Comment (dorzel) [reply internal/pkg/additional/local_stored_collector_test.go:255]: Refactored to use table-driven tests instead

File: internal/pkg/additional/local_stored_collector.go [hunk 5]
```diff
@@ -139,3 +108,79 @@ func (o LocalStorageCollector) AdditionalImagesCollector(ctx context.Context) ([
...
+// buildMirrorToDiskPaths constructs source and destination paths for mirror-to-disk and mirror-to-mirror operations
+func (o LocalStorageCollector) buildMirrorToDiskPaths(img v2alpha1.Image, imgSpec image.ImageSpec, targetRepo, targetTag string) (string, string) {
+	tmpSrc := imgSpec.ReferenceWithTransport
+	var tmpDest string
+
+	if imgSpec.Transport != dockerProtocol {
+		// oci image
+		// Although fetching the digest of the oci image (using o.Manifest.GetDigest) might work in mirrorToDisk and mirrorToMirror
+		// it will not work during diskToMirror as the oci image might not be on the disk any longer
+		tag := "latest"
```
Comment (adolfo-ab) [inline internal/pkg/additional/local_stored_collector.go]: Maybe use a constant for `latest`

File: internal/pkg/additional/local_stored_collector.go [hunk 6]
```diff
@@ -139,3 +108,79 @@ func (o LocalStorageCollector) AdditionalImagesCollector(ctx context.Context) ([
...
+
+	// Docker protocol
+	switch {
+	case imgSpec.IsImageByTagAndDigest():
+		// OCPBUGS-33196 + OCPBUGS-37867- check source image for tag and digest
+		// use tag only for both src and dest
+		o.Log.Warn(collectorPrefix+"%s has both tag and digest : using digest to pull, but tag only for mirroring", imgSpec.Reference)
+		tmpSrc = strings.Join([]string{imgSpec.Domain, imgSpec.PathComponent}, "/") + "@" + imgSpec.Algorithm + ":" + imgSpec.Digest
+		tmpDest = strings.Join([]string{o.destinationRegistry(), targetRepo}, "/") + ":" + targetTag
+	case imgSpec.IsImageByDigestOnly() && img.TargetTag != "":
```
Comment (adolfo-ab) [inline internal/pkg/additional/local_stored_collector.go]: This case produces the same result as default, so we can merge them

File: internal/pkg/additional/local_stored_collector.go [hunk 7]
```diff
@@ -139,3 +108,79 @@ func (o LocalStorageCollector) AdditionalImagesCollector(ctx context.Context) ([
...
+		}
+		tmpSrc := strings.Join([]string{o.LocalStorageFQDN, strings.TrimPrefix(targetRepo, "/")}, "/") + ":" + tag
+		tmpDest := strings.Join([]string{o.Opts.Destination, strings.TrimPrefix(targetRepo, "/")}, "/") + ":" + tag
+		return tmpSrc, tmpDest
+	}
+
+	// Docker protocol
+	var tmpSrc, tmpDest string
+	switch {
+	case imgSpec.IsImageByTagAndDigest():
```
Comment (adolfo-ab) [inline internal/pkg/additional/local_stored_collector.go]: This case is mostly the same as the default , should we merge them?

File: internal/pkg/additional/local_stored_collector.go [hunk 8]
```diff
@@ -139,3 +108,79 @@ func (o LocalStorageCollector) AdditionalImagesCollector(ctx context.Context) ([
...
+	return tmpSrc, tmpDest
+}
+
+// buildDiskToMirrorPaths constructs source and destination paths for disk-to-mirror operations
+func (o LocalStorageCollector) buildDiskToMirrorPaths(img v2alpha1.Image, imgSpec image.ImageSpec, targetRepo, targetTag string) (string, string) {
+	if imgSpec.Transport != dockerProtocol {
+		// oci image
+		tag := "latest"
+		if img.TargetTag != "" {
+			tag = targetTag
```
Comment (adolfo-ab) [inline internal/pkg/additional/local_stored_collector.go:156]: Shouldn't this be 
```suggestion
		if targetTag != "" {
			tag = targetTag
```

File: internal/pkg/additional/local_stored_collector.go [hunk 9]
```diff
@@ -139,3 +108,79 @@ func (o LocalStorageCollector) AdditionalImagesCollector(ctx context.Context) ([
...
+	}
+
+	// Docker protocol
+	var tmpSrc, tmpDest string
+	switch {
+	case imgSpec.IsImageByTagAndDigest():
+		// OCPBUGS-33196 + OCPBUGS-37867- check source image for tag and digest
+		// use tag only for both src and dest
+		o.Log.Warn(collectorPrefix+"%s has both tag and digest : using tag only", imgSpec.Reference)
+		tmpSrc = strings.Join([]string{o.LocalStorageFQDN, targetRepo}, "/") + ":" + targetTag
```
Comment (adolfo-ab) [inline internal/pkg/additional/local_stored_collector.go]: I think using `fmt.Sprintf` for all these paths would be easier to read and less error prone

File: internal/pkg/additional/local_stored_collector.go (hunk 7 above)
Comment (dorzel) [reply internal/pkg/additional/local_stored_collector.go]: Good call, this switch could actually be simplified more too. I'll do that.

File: internal/pkg/additional/local_stored_collector.go [hunk 10]
```diff
@@ -51,68 +51,37 @@ func (o LocalStorageCollector) AdditionalImagesCollector(ctx context.Context) ([
...
+
+		targetTag := imgSpec.Tag
+		if img.TargetTag != "" {
+			targetTag = img.TargetTag
+		}
+
+		switch {
+		case o.Opts.IsMirrorToDisk(), o.Opts.IsMirrorToMirror():
+			tmpSrc, tmpDest = o.buildMirrorToDiskPaths(img, imgSpec, targetRepo, targetTag)
+		case o.Opts.IsDiskToMirror():
```
Comment (r4f4) [inline internal/pkg/additional/local_stored_collector.go]:
```suggestion
		case o.Opts.IsDiskToMirror(), o.Opts.IsDelete():
```
For now, `Delete` is set as the D2M workflow, but in case that changes, let's be explicit here to avoid unexpected breakage.

File: internal/pkg/delete/delete_images.go [hunk 11]
```diff
@@ -166,16 +166,11 @@ func (o DeleteImages) DeleteRegistryImages(deleteImageList v2alpha1.DeleteImageL
...
-		assembleName := name[1] + "/" + imgSpecName.PathComponent
-		// check image type for release or release content
-		switch img.Type {
-		case v2alpha1.TypeOCPReleaseContent:
-			assembleName = name[1] + "/openshift/release"
-		case v2alpha1.TypeOCPRelease:
-			assembleName = name[1] + "/openshift/release-images"
-		}
-		// check the assembled name against the reference name
-		if assembleName != imgSpecRef.Name {
```
Comment (r4f4) [inline internal/pkg/delete/delete_images.go:180]: It's hard to say if this deleted code is necessary or not because there are no unit test cases covering this failure scenario. I think it's better to keep this logic until we introduce unit tests.

File: internal/pkg/delete/delete_images.go [hunk 12]
```diff
@@ -166,16 +166,11 @@ func (o DeleteImages) DeleteRegistryImages(deleteImageList v2alpha1.DeleteImageL
...
+		deleteDestBase := name[1]
+		// Verify that the ImageReference is under the DeleteDestination tree.
+		// This check ensures images are being deleted from the correct registry.
+		// The ImageReference already contains the correct path with targetRepo/targetTag applied.
+		if !strings.HasPrefix(imgSpecRef.Name, deleteDestBase+"/") && imgSpecRef.Name != deleteDestBase {
```
Comment (r4f4) [inline internal/pkg/delete/delete_images.go]: Why are we checking for prefix and equality here? It should be one or the other.

File: internal/pkg/delete/delete_images.go (hunk 11 above)
Comment (dorzel) [reply internal/pkg/delete/delete_images.go:180]: @r4f4 I pushed up some changes for this - the previous code looks to be necessary for the platform type, but we also need the behavior I added for flexibility with types that now support targetRepo/targetTag. This should preserve previous behavior and support the new behavior.

Comment (r4f4) [reply internal/pkg/delete/delete_images.go:180]: Thanks for adding a comment explaining the why.

**Review by aguidirh — APPROVED** [review]
Thanks @dorzel, the PR is working as expected, approving it 🎉 

Test Steps:

With the following ImageSetConfiguration:

```
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v2alpha1
mirror:
  additionalImages:
    - name: registry.redhat.io/ubi9/ubi:latest
      targetRepo: redhat-ubi
      targetTag: ubi999
    - name: registry.redhat.io/ubi8/ubi:latest@sha256:44d75007b39e0e1bbf1bcfd0721245add54c54c3f83903f8926fb4bef6827aa2
    - name: quay.io/prometheus/prometheus:latest 
    - name: registry.redhat.io/ubi8/ubi:latest
    - name: quay.io/rh_ee_aguidi/multi-platform-container:latest
    - name: quay.io/rh_ee_aguidi/empty-image:latest
```

I did `m2d/d2m` and `m2m` and got the following results:

```
Fetching tags for repository: prometheus/prometheus
{
  "name": "prometheus/prometheus",
  "tags": [
    "latest"
  ]
}
Fetching tags for repository: redhat-ubi
{
  "name": "redhat-ubi",
  "tags": [
    "sha256-a3fb564e8be461d5bf8344996eb3ef6eb24a4b8c9333053fe4f3e782657591d3.sig",
    "sha256-c8df11b65b18fcf384d5b7fa53e2a92f381a2d9ad78b9de41dc6f2aa161a0e04.sig",
    "sha256-d263526ff62573b43c6c46c5500fe42f91f88b2ce64c455365852a4d6afba942.sig",
    "sha256-d8660bb38670b46155c2384b4a496547e835bed10c5da4636f313181726c97b4.sig",
    "sha256-fdd18b4d04f5882098bebb3ff057b5b6e239721b113b1e22746cd378ca77e8bf.sig",
    "ubi999"
  ]
}
Fetching tags for repository: rh_ee_aguidi/empty-image
{
  "name": "rh_ee_aguidi/empty-image",
  "tags": [
    "latest",
    "sha256-7d27ba1da0f64410102468d138f9ec8f61f2cc23bb58ff4cd63243a4434e3d99.sig"
  ]
}
Fetching tags for repository: rh_ee_aguidi/multi-platform-container
{
  "name": "rh_ee_aguidi/multi-platform-container",
  "tags": [
    "latest",
    "sha256-02f29c270f30416a266571383098d7b98a49488723087fd917128045bcd1ca75.sig",
    "sha256-0de0e983a4980f32b1aad2d7f7f387cea2d4e9517b47f336cef27f63735911fa.sig",
    "sha256-832f20ad3d7e687c581b0a7d483174901d8bf22bb96c981b3f9da452817a754e.sig",
    "sha256-b15a2f174d803fd5fd7db0b3969c75cee0fe9131e0d8478f8c70ac01a4534869.sig",
    "sha256-e033aa62f84267cf44de611acac2e76bfa4d2f0b6b2b61f1c4fecbefefde7159.sig"
  ]
}
Fetching tags for repository: ubi8/ubi
{
  "name": "ubi8/ubi",
  "tags": [
    "latest",
    "sha256-0a95276776dc776fc42ebaa0ce79aa0b3ac72ab762df1d262beef7b092c2b23e.sig",
    "sha256-0f4503f709fd8f7a2d85049dfd2068f1bf68cd0aae0504bcab677387e455794b.sig",
    "sha256-44d75007b39e0e1bbf1bcfd0721245add54c54c3f83903f8926fb4bef6827aa2.sig",
    "sha256-9bad9c04ec49b48a6ca01b55e7dd6a451010b8b438906af78cb1a1ccc3cb72a3.sig",
    "sha256-f2ae71354f94af134caa8fc6ff9b01bbc974b13c1c783c17927211750bf2c04b.sig"
  ]
}
```

File: internal/pkg/api/v2alpha1/type_config.go (hunk 2 above)
Comment (aguidirh) [reply internal/pkg/api/v2alpha1/type_config.go:292]: This is going to be solved at [CLID-532](https://issues.redhat.com/browse/CLID-532).
